### PR TITLE
[commands] Compound Parser

### PR DESCRIPTION
**Pull request by ikfir**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation)
- [x] 🐞 Bug fix (fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (adds functionality)

### 📚 Description
Compound parser is a simple way to create custom parsers that are composed of other parsers
For example if we want to create a simple location parser it will be the following:
```java
@WithParser(DoubleParser.class)
@WithParser(FloatParser.class)
@WithParser(WorldParser.class)
public class LocationParser extends CompoundParser<Location> {
    
    public static final String DEFAULT_KEYWORD = "location";
    
    public LocationParser(int priority, String keyword) {
        super(
          keyword,
          Location.class,
          priority,
          new SimpleArgumentMapper(),
          new SimpleCommandLexer()
        );
    }

    public LocationParser(int priority) {
        this(priority, DEFAULT_KEYWORD);
    }

    @ParserVariant("<world> <double> <double> <double>")
    public Location parseWorldWithXyz(World world, double x, double y, double z) {
        return new Location(world, x, y, z);
    }


    @ParserVariant("<world> <double> <double> <double> <float> <float>")
    public Location parseWorldWithXyzYawPitch(World world, double x, double y, double z, float yaw, float pitch) {
        return new Location(world, x, y, z, yaw, pitch);
    }
    

    @SenderLimit(SenderType.PLAYER)
    @ParserVariant("<double> <double> <double>")
    public Location parseWithXyz(Player sender, double x, double y, double z) {
        return new Location(
                sender.getWorld(),
                x,
                y,
                z
        );
    }

    @SenderLimit(SenderType.PLAYER)
    @ParserVariant("<double> <double> <double> <float> <float>")
    public Location parseWithXyzYawPitch(Player sender, double x, double y, double z, float yaw, float pitch) {
        return new Location(
                sender.getWorld(),
                x,
                y,
                z,
                yaw,
                pitch
        );
    }
}
```

### 🧪 How Has This Been Tested?
Unit testing

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

## Release Notes

- **New Features**
  - Introduced a "Compound Parser" for enhanced command parsing capabilities.
  - Added support for named command arguments and improved command handling.
  - New `CatCommand` and `Meow` parser to manage cat-related commands.
  - Added performance testing for command execution.
  - Introduced `PositionParser` for parsing position-related commands.
  - Enhanced command requirement handling with a new `GenericNode` interface.
  - Added a `PermissionCommand` for managing permissions in commands.
  - New `ParserVariant` and `ParserVariants` annotations for flexible command parsing.
  - Added `AnotherEvilCompoundParser` and `EvilCompoundParser` for specialized command handling.
  - Enhanced `PermissionFactory` for improved permission handling.

- **Documentation**
  - Updated changelog and added new documentation for the "Compound Parser."
  - Enhanced documentation for various components and methods, including the `PositionParser`.

- **Bug Fixes**
  - Improved error handling and reporting in command tests.
  - Adjusted method signatures for broader input compatibility.

- **Tests**
  - Added unit tests for new parser functionalities and command execution performance.
  - Introduced tests for the `PositionParser` to validate parsing and tab completion.
  - Added performance benchmark tests for command processing speed.
  - Added tests for the `PermissionCommand` functionality.
  - Added tests for the `PermissionFactory` to ensure null handling.
  - Added tests for `RegisteredVariant` equality logic validation.
  - Added tests for the `EvilCompoundParser` and `AnotherEvilCompoundParser` instantiation.

### CODE REVIEW

File: commands/src/test/java/net/apartium/cocoabeans/commands/parsers/compound/PositionParser.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public Position pos(double x, double y, double z) {
+        return new Position(x, y, z);
+    }
+
+    @WithParser(DoubleParser.class)
+    @WithParser(value = IntParser.class, priority = 1)
+    @ParserVariant("<int> <double> <int>")
+    public Position pos(int x, double y, int z) {
+        return new Position(x * 5, y, z);
+    }
```
[the remark this answers is not in the document]
Comment (ikfir) [reply commands/src/test/java/net/apartium/cocoabeans/commands/parsers/compound/PositionParser.java:31]: just a test doesn't really need it

File: commands/spigot-platform/src/test/java/net/apartium/cocoabeans/commands/spigot/parsers/MeowParser.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+
+    public MeowParser(int priority) {
+        super("meow", Meow.class, priority, new SimpleArgumentMapper(), new SimpleCommandLexer());
+    }
+
+    @ParserVariant("<string> <int> <gender>")
+    public Meow serialize(String cat, int age, Meow.Gender gender) {
+        return new Meow(cat, age, gender);
+    }
```
[the remark this answers is not in the document]
Comment (ikfir) [reply commands/spigot-platform/src/test/java/net/apartium/cocoabeans/commands/spigot/parsers/MeowParser.java:23]: test doesn't mattar

File: commands/src/test/java/net/apartium/cocoabeans/commands/GeneralCommandTest.java [hunk 3]
```diff
@@ -345,6 +345,19 @@ void yesMeow() {
...
+        long start, end;
+
+        for (int i = 0; i < 16_384; i++) {
+            start = System.currentTimeMillis();
+            evaluate("test", "try random-string 0.3");
+            end = System.currentTimeMillis();
+            assertEquals(List.of("tryDouble(Sender sender, double num) I ignore the second argument it wasn't important also your number is 0.3"), sender.getMessages());
+            assertTrue(end - start < 50);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ikfir) [reply commands/src/test/java/net/apartium/cocoabeans/commands/GeneralCommandTest.java:359]: We may need to open issue about stress test

File: Writerside/topics/Examples.md [hunk 4]
```diff
@@ -14,4 +14,5 @@ You can find some examples of using the commands system in this page
 * [`Write information about command`](Command-Information.md)
 * [`Dynamic annotation parser`](Dynamic-annotation-parser.md)
 * [`Value Requirement`](value-requirement.md)
-* [`Named command arguments`](parameter-name.md)
\ No newline at end of file
+* [`Named command arguments`](parameter-name.md)
+* [`Compound parser`]()
```
Comment (liorsl) [inline Writerside/topics/Examples.md:18]: need link here

File: commands/spigot-platform/src/main/java/net/apartium/cocoabeans/commands/spigot/requirements/factory/PermissionFactory.java [hunk 5]
```diff
@@ -51,6 +56,22 @@ public RequirementResult meetsRequirement(RequirementEvaluationContext context)
             return RequirementResult.meet();
         }
 
+        private RequirementResult meetsInvertedRequirement(RequirementEvaluationContext context) {
+            Sender sender = context.sender();
+
+            if (sender.getSender() == null || !(sender.getSender() instanceof CommandSender commandSender))
+                return RequirementResult.meet();
```
Comment (liorsl) [inline commands/spigot-platform/src/main/java/net/apartium/cocoabeans/commands/spigot/requirements/factory/PermissionFactory.java:63]: shouldn't it fail? even if its inverted, its not a valid sender

**Review by liorsl — CHANGES_REQUESTED** [review]

**Review by liorsl — APPROVED** [review]

**Pull request merged**